Stack/Opc.Ua.Core/Stack/Tcp/TcpTransportListener.cs

```diff
@@ -37,6 +38,233 @@ public override ITransportListener Create()
+        }
+
+        /// <summary>
+        /// Adds an action entry for a potential rogue client
+        /// </summary>
+        /// <param name="ipAddress"></param>
+        public void AddRogueClientAction(IPAddress ipAddress)
+        {
+            int currentTicks = HiResClock.TickCount;
+
+            m_rogueClients.AddOrUpdate(ipAddress,
+                // If client is new , create a new entry
+                key => new RogueClient {
+                    LastRogueActionTicks = currentTicks,
+                    RogueActionCount = 1,
+                    BlockedUntilTicks = 0
+                },
+                // If the client exists, update its entry
+                (key, existingEntry) => {
+                    // If IP currently blocked simply do nothing
+                    if (IsBlockedTicks(existingEntry.BlockedUntilTicks, currentTicks))
+                    {
+                        return existingEntry;
+                    }
+
+                    // Elapsed time since last recorded action
+                    int elapsedSinceLastRecAction = currentTicks - existingEntry.LastRogueActionTicks;
+
+                    if (elapsedSinceLastRecAction <= m_kRogueIntervalMs)
+                    {
+                        existingEntry.RogueActionCount++;
+
+                        if (existingEntry.RogueActionCount > m_kNrActionsTillBlock)
+                        {
+                            // Block the IP
+                            existingEntry.BlockedUntilTicks = currentTicks + m_kBlockDurationMs;
+                            Utils.LogError("RemoteClient IPAddress: {0} blocked for {1} ms due to exceeding {2} rogue actions under {3} ms ",
+                                ipAddress.ToString(),
+                                m_kBlockDurationMs,
+                                m_kNrActionsTillBlock,
+                                m_kRogueIntervalMs);
+
+                        }
+                    }
+                    else
+                    {
+                        // Reset the count as the last action was outside the interval
+                        existingEntry.RogueActionCount = 1;
+                    }
+
+                    existingEntry.LastRogueActionTicks = currentTicks;
+
+                    return existingEntry;
+                }
+            );
+        }
+
+        /// <summary>
+        /// Dispose the cleanup timer
+        /// </summary>
+        public void Dispose()
+        {
+            m_cleanupTimer?.Dispose();
+        }
+
+        #endregion
+        #region Private methods
+
+        /// <summary>
+        /// Periodically cleans up expired RogueClientEntries to avoid memory leak and unblock clients whose duration has expired.
+        /// </summary>
+        /// <param name="state"></param>
+        private void CleanupExpiredEntries(object state)
+        {
+            int currentTicks = HiResClock.TickCount;
+
+            foreach (var entry in m_rogueClients)
```

Comment:
m_rogueClients.ToArray first to get all entries as iterating over a changing concurrent dictionary could miss values ( I think so at least)

Reply:
Even if this would be true since this is the clean-up thread I think this is just going to miss the newly added entries which will not be subject to removal anyhow ...